include/boost/beast/core/buffers_generator.hpp

```diff
@@ -0,0 +1,229 @@
+    void operator()(Self& self,
+                    error_code ec = {},
+                    std::size_t n = 0)
+    {
+        BOOST_ASIO_CORO_REENTER(*this)
+        {
+            for(;;)
+            {
+                BOOST_ASIO_CORO_YIELD
+                {
+                    auto cb = generator_.prepare(ec);
+                    if (ec)
+                        goto complete;
+                    if (boost::beast::buffer_bytes(cb) == 0)
+                        goto complete;
+                    stream_.async_write_some(
+                        cb, std::move(self));
+                }
+                generator_.consume(n);
+
+                total_ += n;
+            }
+
+        complete:
+            self.complete(ec, total_);
+        }
+    }
+
+private:
+    AsyncWriteStream& stream_;
+    BuffersGenerator generator_;
+    std::size_t total_ = 0;
+};
+
+} // detail
+
+//----------------------------------------------------------
+
+template <
+    class SyncWriteStream,
+    class BuffersGenerator
+#if ! BOOST_BEAST_DOXYGEN
+    ,
+    typename std::enable_if<
+        is_buffers_generator<typename std::decay<
+            BuffersGenerator>::type>::value>::type* =
+        nullptr
+#endif
+    >
+size_t
+write(SyncWriteStream& stream,
+      BuffersGenerator&& generator,
+      beast::error_code& ec)
+{
+    static_assert(
+        is_sync_write_stream<SyncWriteStream>::value,
+        "SyncWriteStream type requirements not met");
+
+    ec.clear();
+    size_t total = 0;
+    for (;;) {
+        auto cb = generator.prepare(ec);
+        if (ec)
+            break;
+        if (boost::beast::buffer_bytes(cb) == 0)
+            break;
+
+        size_t n = net::write(stream, cb, ec);
+
+        if (ec)
+            break;
+
+        generator.consume(n);
+        total += n;
+    }
+
+    return total;
+}
+
+//----------------------------------------------------------
+
+template <
+    class SyncWriteStream,
+    class BuffersGenerator
+#if ! BOOST_BEAST_DOXYGEN
+    ,
+    typename std::enable_if<
+        is_buffers_generator<typename std::decay<
+            BuffersGenerator>::type>::value>::type* =
+        nullptr
+#endif
+    >
+std::size_t
+write(SyncWriteStream& stream, BuffersGenerator&& generator)
+{
+    static_assert(
+        is_sync_write_stream<SyncWriteStream>::value,
+        "SyncWriteStream type requirements not met");
+    beast::error_code ec;
+    std::size_t n = write(
+        stream, std::forward<BuffersGenerator>(generator),
+        ec);
+    if (ec)
+        BOOST_THROW_EXCEPTION(system_error{ ec });
+    return n;
+}
+
+//----------------------------------------------------------
+
+template <
+    class AsyncWriteStream,
+    class BuffersGenerator,
+    class CompletionToken
+#if !BOOST_BEAST_DOXYGEN
+    ,
+    typename std::enable_if<
+        is_buffers_generator<BuffersGenerator>::value //
+        >::type* = nullptr
+#endif
+    >
+auto
+async_write(AsyncWriteStream& stream,
```

Comment:
This needs to go in `<boost/beast/core/impl/buffers_generator.hpp>` so we don't force the doc toolchain to parse it needlessly.